### ISSUE TITLE
[Feature Request/Discussion] Runtime "Theme Engine" aka advanced tinting

### ISSUE THREAD

**Issue by MTRNord** [vector-im/riot-web#5844]
I asked yesterday in #riot-dev what the opinions about allowing to load css at runtime via a UI plus a Theme design UI or advanced tinting aka. allowing to modify different components by CSS values easily. (You can reread it here: https://matrix.to/#/!DdJkzRliezrwpNebLk:matrix.org/$15133678091603OtUhP:matrix.ffslfl.net )

This issue is more a Request/Tracking issue about what we already discussed and what the Community prefers.

Things we discussed:
- Similiar possibilities to what Telegram in their clients allows
- Allowing to modify Components like Buttons independently from the Rest of the Theme
- Allowing to save these Themes in a file format (Propably a css file is not wanted as you could abuse it. I am thinking about a XML file that gets parsed)
- Having some UI for generating these Theme files
- In general making custom Themes for non Devs easier. Especialy if someone uses the Desktop App ( https://matrix.to/#/!DdJkzRliezrwpNebLk:matrix.org/$15133685521608ayXMf:matrix.ffslfl.net )

**Comment by pafcu** [comment]
I'm not saying it's not possible, but how could the CSS be abused, and would it be possible to prevent those things specifically?

**Comment by MTRNord** [comment]
@pafcu depending on how you load it you could do `</style><script>something();</script>` but of course you might normalize it to strip these things anyway.

**Comment by pafcu** [comment]
I'd just add the content of the CSS file as text to a script element. That automatically prevents that.

**Comment by MTRNord** [comment]
@pafcu well thats something I didn't know of before :) In that case a CSS would work of course

**Comment by pafcu** [comment]
In any case, you could probably do a bit of damage with CSS anyway (think fake display names, hide encryption warnings, leak usage by loading external resources) so we need to consider how much we want to trust theme authors, and what we want to prevent / sanitize.

**Comment by MTRNord** [comment]
@pafcu I feel like that the `color:` value only should allow colors that aren't near the background color for example. Means to check their value if it is similiar or not

**Comment by MTRNord** [comment]
Also using custom svg might make problems with plain css files.

**Comment by pafcu** [comment]
What problems would custom SVG cause?

**Comment by MTRNord** [comment]
@pafcu can you replace the svg files of the buttons with plain css? aren't thoose defined via html objects?

**Comment by pafcu** [comment]
But that applies to most images, not just SVG (unless they are implemented using CSS bsckground, which causes other problems)

**Comment by lampholder** [comment]
How far is Slack's impl from what you're looking for?

Slack supports a very limited custom styling, enumerating stylable components and allowing you to specify any colour value:
![image](https://user-images.githubusercontent.com/1922197/34113559-ffaaf792-e407-11e7-9de1-8bb9ef04e149.png)

And it has no defense against a fully auberginetastic theme, though it is purely per-client (you share themes by users' copy+pastaing color value strings into their own client):
![image](https://user-images.githubusercontent.com/1922197/34113620-36766c7a-e408-11e7-9ec5-84d29b7d3842.png)

**Comment by MTRNord** [comment]
What Telegram does: Atleast on mobile: 
https://telegram.org/blog/android-themes

**Comment by turt2live** [comment]
I'd be kinda surprised if Riot did actually restrict colors, as it'd probably get in the way more than normal or be too inaccurate (one person's definition of "close to" is going to differ from another's). If the user really wants everything to be hot pink, let them have it at their own risk. 

I like slack's approach, and I've seen it quite often outside of chat as well. Riot could do with a similar number of options where you won't be able to change things like the encryption status icons.

**Comment by pafcu** [comment]
I'm honestly not a huge fan of the Slack syntax for it, because just having a list like that much be terrible for backward compatibility. Perhaps e.g. Base64-encoded compressed JSON would be able to achieve something similar?

**Comment by MTRNord** [comment]
I agree with the UI of Slack but as @pafcu says not with the Syntax/format of data

**Comment by turt2live** [comment]
Keeping it in an easily editable form makes sense to me. Base64 is probably going over the top, but JSON wouldn't be a bad idea. Realistically it should be something that someone can paste into the box and be able to edit it *in the field* - and know what they're changing.

**Comment by pafcu** [comment]
I don't see why you'd want to be able to edit by hand when you have a perfectly fine UI to do it.

**Comment by turt2live** [comment]
People may want to put the thing in version control, company documentation, or really anywhere that isn't Riot. It's relatively common for companies here to provide that information at least in internal documentation. It's more so that marketing and ops (or whichever team is deploying the thing) can keep things consistent, even if both departments go through a complete turnover.

**Comment by pafcu** [comment]
Hmm, perhaps accept both? It seems like for your case it would anyway be much better to have themes in a separate file, rather than a copy-pastable string.

**Comment by MTRNord** [comment]
I agree with the file approach, Makes it simpler (and is what Telegram for example does)

**Comment by turt2live** [comment]
I just don't see the benefit of having a "share with your friends" section if it's hard to share it with your friends. 

Consider the following scenario:
* Bob comes up with a theme that they think looks nice
* Bob sends the theme to Alice
* Alice likes it, but changes a couple values
* Alice sends it back to Bob

If Bob can't see Alice's changes easily, there's no real point in that exchange at all, making it become one-sided (Alice tells Bob to change the colors instead of doing it themselves). 

A file might work, but I really don't see a problem with a JSON string. It's not going to be that long.

**Comment by turt2live** [comment]
RE-reading the original post does mean the JSON would be insanely long, as it implies being able to change every minuscule aspect of the app. From experience, this is a somewhat bad idea as it leads to confusion and very complex UIs. Slack's approach, where you're only able to edit 8 things, seems like a better way forward: it reduces the complexity of the UI, is approachable, and keeps the app looking at least somewhat reasonable.

**Comment by pafcu** [comment]
> I just don't see the benefit of having a "share with your friends" section if it's hard to share it with your friends.

I _assumed_ that is why Slack has an easily copy-pasteable string, but it might of course be for some completely other reason.

**Comment by MTRNord** [comment]
@turt2live well you can just load the json into a file. That makes it again easier to have them in repos. :) just download the raw file and import it or maybe even easier (for a user) to just use the file link. (but harder to implement)

**Comment by turt2live** [comment]
A file is a bit harder to deal with when it's embedded in internal documentation though. Really it ends up being how long this string is. If Riot allows the user to change every tiny aspect then I'd recommend a file. If it's only 8 colors, then I'd recommend it just be a string - the people who want it in repos can put it in a file if they want to.

**Comment by turt2live** [comment]
As another data point, here's a fairly rough UI I ended up creating for my day job. It's a proof of concept for theming the application so that customers can re-sell the product.

![image](https://user-images.githubusercontent.com/1190097/34118039-d5b585c4-e3da-11e7-9cfb-9ff5b2acfbc2.png)

**Comment by ara4n** [comment]
I think we should provide a basic colour picker and font picker to let people easily reskin the app a bit for their personal preferences in terms of contrast, font legibility (particularly proportional v. monospace, and supporting Greek & Japanese etc to mitigate like https://github.com/vector-im/riot-web/issues/7949).  Brownie points if we can do it per-room via granular settings so we can recover the old tinting functionality.

**Comment by Austcool-Walker** [comment]
I'm more concerned with the loading of a custom font installed on the OS with riot tbh. For example on Discord and other chat clients on linux can be forced to load a custom font OS wide if a font.conf (.xml) is placed in `~/.config/fontconfig/fonts.conf` matrix is electron like discord but for some reason it ignores/overrides the font config file in the home folder.

**Comment by ara4n** [comment]
Just in case anyone missed it, we do now have a runtime theme engine as per https://github.com/vector-im/riot-web/blob/develop/docs/theming.md.  It needs to be exposed nicely, though. https://github.com/matrix-org/matrix-react-sdk/pull/4148 could be one step in the right dir.

**Comment by MTRNord** [comment]
Thats awesome :D That is 60% of what I had in mind :) (I was more on the telegram route. But having that ability for hosters is a good starting point)
